### ISSUE TITLE
Test Chaste compatibility with newer Ubuntus

### ISSUE THREAD

**Issue by fcooper8472** [Chaste/Chaste#198]
- [x] Ubuntu 22.10 Kinetic Kudu
- [x] Ubuntu 23.04 Lunar Lobster
- [x] Ubuntu 23.10 Mantic Minotaur

**Comment by fcooper8472** [comment]
### Ubuntu 22.10

Continuous and Nightly test packs compile and pass with no changes required. The following configuration was installed with the chaste-dependencies package:

```xml
6: <ChasteBuildInfo>
6:      <ProvenanceInfo>
6:              <VersionString>2021.1.71123d1</VersionString> <!-- build specific -->
6:              <IsWorkingCopyModified>0</IsWorkingCopyModified>
6:              <BuildInformation>Debug</BuildInformation>
6:              <BuildTime>Tue, 05 Dec 2023 21:16:10 +0000</BuildTime>
6:              <CurrentTime>Tue, 05 Dec 2023 21:52:52 +0000</CurrentTime>
6:              <BuilderUnameInfo>Linux</BuilderUnameInfo>
6:              <ChasteTestOutputDirectory>/home/fcooper/GitRepos/Chaste/chaste-test-dir/</ChasteTestOutputDirectory>
6:              <Projects>
6:              </Projects>
6:      </ProvenanceInfo>
6:      <Compiler>
6:              <NameAndVersion>/usr/bin/c++, version 12.2.0</NameAndVersion>
6:              <Flags>-Wall -Werror -Wnon-virtual-dtor -Woverloaded-virtual -Wextra -Wno-attributes -Wno-unused-parameter -Wvla -Wimplicit-fallthrough=2 -g</Flags>
6:      </Compiler>
6:      <Libraries>
6:              <CompiledIn>
6:                      <PETSc>3.16.6</PETSc>
6:                      <Boost>1.74.0</Boost>
6:                      <HDF5>1.10.8</HDF5>
6:                      <Parmetis>4.0.3</Parmetis>
6:              </CompiledIn>
6:              <Binaries>
6:                      <XSD>4.0.0</XSD>
6:              </Binaries>
6:              <Optional>
6:                      <SUNDIALS>5.8.0</SUNDIALS>
6:                      <VTK>7.1</VTK>
6:                      <Xerces>3.2.3</Xerces>
6:              </Optional>
6:      </Libraries>
6:      <ChasteCodegenVersion>chaste_codegen 0.10.2</ChasteCodegenVersion>
6: </ChasteBuildInfo>
```

**Comment by fcooper8472** [comment]
### Ubuntu 23.04

This currently fails due to the following requirement in the Debian control file:

```
libvtk7-dev | libvtk6-dev
```

This version of Ubuntu dropped libvtk 6, 7 AND 8: it looks as though the only vtk available is 9.

- [x] Recommendation: update control file to `libvtk9-dev`, as this is the only vtk available via the package manager on Ubuntus 23.04, 23.10 and (according to info available so far) 24.04 LTS.

**Comment by fcooper8472** [comment]
### Other updates to tidy control file

- [x] Retire the `scons` part of `cmake | scons`
- [x] Update `libpetsc-real3.15-dbg|libpetsc-real3.14-dbg|libpetsc-real3.12-dbg` to `libpetsc-real3.18-dbg` (3.18 appears to be in use for 23.04, 23.10 and 24.04)
- [x] Stop recommending `libfltk` (we're not using it, are we?!)
- [x] Stop suggesting `libgoogle-perftools-dev` as we've even dropped it from our CI because we couldn't get it to work
- [x] Stop suggesting `subversion, git-svn, gnuplot`: as far as I know we're not using any of them?
- [x] Move remaining suggestions to recommendations, so we can get rid of the horrible instructions for getting suggested packages

@mirams and @jmpf do you have any comments on any of these proposed changes?

**Comment by mirams** [comment]
Keep gnuplot! Used to use that all the time when debugging numerics...

**Comment by jmpf** [comment]
Surely people who need gnuplot know how to install it?

**Comment by fcooper8472** [comment]
I have updated the control file as discussed. Full details here:
https://github.com/Chaste/infrastructure-scripts/pull/1

I have installed Ubuntus 22.10, 23.04, 23.10 and compiled and run the tests. Everything is fine, except:

1. 23.04 and 23.10 fail the citation test due to an updated PETSc citation
2. The builds fail due to c_vector warnings (see #197), so currently only compile when warnings are not treated as errors

**Comment by bdevans** [comment]
> ### Other updates to tidy control file
> * [x]  Retire the `scons` part of `cmake | scons`
> * [x]  Update `libpetsc-real3.15-dbg|libpetsc-real3.14-dbg|libpetsc-real3.12-dbg` to `libpetsc-real3.18-dbg` (3.18 appears to be in use for 23.04, 23.10 and 24.04)
> * [x]  Stop recommending `libfltk` (we're not using it, are we?!)
> * [x]  Stop suggesting `libgoogle-perftools-dev` as we've even dropped it from our CI because we couldn't get it to work
> * [x]  Stop suggesting `subversion, git-svn, gnuplot`: as far as I know we're not using any of them?
> * [x]  Move remaining suggestions to recommendations, so we can get rid of the horrible instructions for getting suggested packages
> 
> @mirams and @jmpf do you have any comments on any of these proposed changes?

- ~[ ] Check whether the `chaste-dependencies` should have `python3-dev` (as I found it needed at some distant point) or just `python3` as it currently is.~ On `noble`, this no longer seems to be necessary (at least I get the same test failures with it as without).
- ~[ ] Add `python3-pip` to dependencies since I think it is needed to install `chaste-codegen` (although this might get pulled in by venv).~ This is pulled in and managed by `venv`.

**Comment by bdevans** [comment]
When building the `chaste/develop` docker image on `noble`, the following tests fail:
```
The following tests FAILED:
          3 - TestCitations (Failed)
         78 - TestQuadraticMesh (Failed)
         93 - TestVtkMeshReader (Subprocess aborted)
        113 - TestXmlMeshWriters (Failed)
        175 - TestHdf5Converters (Subprocess aborted)
        190 - TestCompressibleNonlinearElasticitySolver (Failed)
        198 - TestStokesFlowSolver (Failed)
        262 - TestNodeBasedCellPopulation (Failed)
        290 - TestOnLatticeSimulationWithPottsBasedCellPopulation (Failed)
        367 - TestMajorAirwaysCentreLinesCleaner (Subprocess aborted)
        382 - TestBidomainProblem (Failed)
        420 - TestMonodomainProblem (Failed)
        432 - TestPostProcessingWriter (Subprocess aborted)
```

**Comment by fcooper8472** [comment]
I think we can probably expect noble to be pretty unstable for the time being, so not necessarily the case that all of those are things we will need to address (we still haven't even reached the kernel freeze).

Some are definitely "real", e.g. TestCitations is because PETSc now emits a new citation.

**Comment by bdevans** [comment]
Fair enough, just following up after reporting a bug in #238 and wanted to see if it was present on `noble` so thought it was worth giving a heads up on the current state of that build.

**Comment by bdevans** [comment]
Does anyone know if `texttest` is still used? If so, I believe the variable `TEXTTEST_HOME=/usr/local/bin/texttest` needs to be updated now that it needs to be installed in a `venv` - should it just be set to where the `venv` is with the following commands (i.e. `TEXTTEST_HOME="${CHASTE_BUILD_DIR}/texttest_venv"`)?
```
python -m venv --upgrade-deps "${CHASTE_BUILD_DIR}/texttest_venv" && \
. "${CHASTE_BUILD_DIR}/texttest_venv/bin/activate" && \
python -m pip install texttest
```

**Comment by fcooper8472** [comment]
> Does anyone know if `testtext` is still used? If so, I believe the variable `TEXTTEST_HOME=/usr/local/bin/texttest` needs to be updated now that it needs to be installed in a `venv` - should it just be set to where the `venv` is with the following commands (i.e. `TEXTTEST_HOME="${CHASTE_BUILD_DIR}/testtext_venv"`)?
> 
> ```
> python -m venv --upgrade-deps "${CHASTE_BUILD_DIR}/testtext_venv" && \
> . "${CHASTE_BUILD_DIR}/testtext_venv/bin/activate" && \
> python -m pip install texttest
> ```

I'll ping @mirams  and @jmpf for this

**Comment by jmpf** [comment]
> > Does anyone know if `testtext` is still used? If so, I believe the variable `TEXTTEST_HOME=/usr/local/bin/texttest` needs to be updated now that it needs to be installed in a `venv` - should it just be set to where the `venv` is with the following commands (i.e. `TEXTTEST_HOME="${CHASTE_BUILD_DIR}/testtext_venv"`)?
> > ```
> > python -m venv --upgrade-deps "${CHASTE_BUILD_DIR}/testtext_venv" && \
> > . "${CHASTE_BUILD_DIR}/testtext_venv/bin/activate" && \
> > python -m pip install texttest
> > ```
> 
> I'll ping @mirams and @jmpf for this

texttest is/was used for acceptance testing.  See #244

**Comment by ptheywood** [comment]
> TestPostProcessingWriter



> When building the `chaste/develop` docker image on `noble`, the following tests fail:
> 
> ```
> The following tests FAILED:
>           3 - TestCitations (Failed)
>          78 - TestQuadraticMesh (Failed)
>          93 - TestVtkMeshReader (Subprocess aborted)
>         113 - TestXmlMeshWriters (Failed)
>         175 - TestHdf5Converters (Subprocess aborted)
>         190 - TestCompressibleNonlinearElasticitySolver (Failed)
>         198 - TestStokesFlowSolver (Failed)
>         262 - TestNodeBasedCellPopulation (Failed)
>         290 - TestOnLatticeSimulationWithPottsBasedCellPopulation (Failed)
>         367 - TestMajorAirwaysCentreLinesCleaner (Subprocess aborted)
>         382 - TestBidomainProblem (Failed)
>         420 - TestMonodomainProblem (Failed)
>         432 - TestPostProcessingWriter (Subprocess aborted)
> ```

This includes all the tests which fail on 22.04 due to  due to parsing `.vtu` files that contain binary data - #249 (everything other than `TestCitations`, `TestStokesFlowSolver`, `TestNodeBasedCellPopulation`). 
Unfortunately I don't expect downgrading libexpat1 via apt will be an option here as it only appears to have as single build available.

This has now been fixed upstream by VTK, but no indication of when this will make it into Ubuntu repositories / backports.

**Comment by mirams** [comment]
See also #249 and #257

**Comment by fcooper8472** [comment]
Closed by #197 

Chaste now works with all Ubuntu versions up to and including 24.04.
